core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -631,14 +632,19 @@ && isTaskScaleWritersEnabled(session)
                         newSource.getProperties());
```

Comment:
nit: Could we split `visitPartitionedWriter` into two methods (part and unpart). It's odd to see `visitPartitionedWriter` called with `Optional.empty()`